src/Authenticator/SessionAuthenticator.php

```
@@ -115,4 +117,80 @@ public function clearIdentity(ServerRequestInterface $request, ResponseInterface
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request
+     * @param \Psr\Http\Message\ResponseInterface $response The response
+     * @param \ArrayAccess $impersonator User who impersonates
+     * @param \ArrayAccess $impersonated User impersonated
+     * @return array
+     */
+    public function impersonate(
+        ServerRequestInterface $request,
+        ResponseInterface $response,
+        \ArrayAccess $impersonator,
+        \ArrayAccess $impersonated
+    ): array {
+        $sessionKey = $this->getConfig('sessionKey');
+        $impersonateSessionKey = $this->getConfig('impersonateSessionKey');
+        /** @var \Cake\Http\Session $session */
+        $session = $request->getAttribute('session');
+        if ($session->check($impersonateSessionKey)) {
+            throw new UnauthorizedException(
+                'You are impersonating a user already. Stops impersonation before doing another one.'
+            );
+        }
+        $session->write($impersonateSessionKey, $impersonator);
+        $session->write($sessionKey, $impersonated);
+        $this->setConfig('identify', true);
+
+        return [
+            'request' => $request,
+            'response' => $response,
+        ];
+    }
+
+    /**
+     * Stops impersonation
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request
+     * @param \Psr\Http\Message\ResponseInterface $response The response
+     * @return array
+     */
+    public function stopImpersonating(ServerRequestInterface $request, ResponseInterface $response): array
+    {
+        $sessionKey = $this->getConfig('sessionKey');
+        $impersonateSessionKey = $this->getConfig('impersonateSessionKey');
+        /** @var \Cake\Http\Session $session */
+        $session = $request->getAttribute('session');
+        if (!$session->check($impersonateSessionKey)) {
+            throw new UnauthorizedException('You are not impersonating a user.');
```

Should this be silent error? Won't raising an exception here cause the request to fail?

---

it has been fixed in latest commit. I will mark it as resolved but feel free to reopen to discuss anything else about this.

```php
/** @var \Cake\Http\Session $session */
$session = $request->getAttribute('session');
if ($session->check($impersonateSessionKey)) {
    $identity = $session->read($impersonateSessionKey);
    $session->delete($impersonateSessionKey);
    $session->write($sessionKey, $identity);
    $this->setConfig('identify', true);
}
```